tritium-metrics/src/main/java/com/palantir/tritium/metrics/TaggedMetricRegistry.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.ExponentiallyDecayingReservoir;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * Similar to {@link MetricRegistry} but allows tagging of {@link Metric}s.
+ */
+public final class TaggedMetricRegistry {
+
+    private static final TaggedMetricRegistry DEFAULT = new TaggedMetricRegistry();
+
+    private final Map<MetricName, Metric> registry = new ConcurrentHashMap<>();
+
+    public TaggedMetricRegistry() {}
+
+    /**
+     * Get the global default {@link TaggedMetricRegistry}.
+     */
+    public TaggedMetricRegistry getDefault() {
```

Comment:
Is the assumption that metric producers will register via this globally accessible singleton registry? I'm not opposed to this as plumbing the old `MetricRegistry` down through many layers is often a challenge. Might want to consider testability and how folks should handle this to prevent poor concurrent unit test interactions.

Reply:
so one idea was for classes to grab a metric registry the same way classes grab a logger:

    private static final TaggedMetricRegistry registry = 
        MetricContext.getTaggedRegistry(ClassName.class);

Reply:
might then allow for tests to adjust the `MetricContext` in some smart way.

Reply:
think we should only do that if we think sticking the FQN on all metrics is useful

Reply:
I'm assuming that would pre-populate any metrics with a tag like `class: ClassName` (or similar), is that correct? How does that play into API where you provide the `MetricName` that can already include tags, and which one trumps in case of tag name collision?

Reply:
I thought one of the tags was a class name. The other way to think about it is that this is a way to get the 'right' registry but to still allow it to be statically grabbed.

Reply:
the list of tags we've discussed doesn't include class name. I'm also not particularly concerned with a single global registry for now (given we only use a single instance of `MetricRegistry` everywhere right now). If we think class name is a useful tag in general then happy to add the `get(Class)` method

Reply:
Actually this method needs to be `static`, correct?